File: dynamic_programming/book_shop/book_shop.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Knapsack problem (good to review)
int max_pages(vector<int> prices, vector<int> pages, int max_price) {
    vector<int> dp(max_price + 1);
    for (int i = 0; i < prices.size(); i++) {
        int price = prices[i];
        for (int j = max_price - price; j >= 0; j--) {
            dp[j + price] = max(
                dp[j + price], dp[j] + pages[i]
            );
        }
    }
    return dp[max_price];
}

vector<int> read_nums(int size) {
    vector<int> nums(size);
    for (int i = 0; i < size; i++) {
        int num;
        scanf("%d", &num);
        nums.push_back(num);
    }
    return nums;
}

int read_num() {
    int num;
    scanf("%d", &num);
    return num;
}

int main() {
    int num_books = read_num();
    int max_price = read_num();
    vector<int> prices = read_nums(num_books);
    vector<int> pages = read_nums(num_books);

    cout << max_pages(prices, pages, max_price) << endl;
}

File: dynamic_programming/book_shop/book_shop.py
from sys import stdin

# TLE 
def book_shop(prices, pages, max_price):
    dp = [0 for _ in range(max_price + 1)]
    for i in range(len(prices)):
        price = prices[i]
        for x in range(max_price - price, -1, -1):
            dp[x + price] = max(dp[x] + pages[i], dp[x + price])
    return dp[max_price]

def read_num():
    return int(stdin.readline())

def read_nums():
    return [int(num) for num in stdin.readline().split()]

if __name__ == '__main__':
    _, max_price = read_nums()
    prices = read_nums()
    pages = read_nums()
    print(book_shop(prices, pages, max_price))
    